src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_windows.cc

```diff
@@ -193,74 +199,86 @@ class GrpcPolledFdWindows {
     grpc_socket_notify_on_read(winsocket_, &outer_read_closure_);
   }
 
-  void RegisterForOnWriteableLocked(grpc_closure* write_closure) {
-    if (socket_type_ == SOCK_DGRAM) {
-      GRPC_CARES_TRACE_LOG("fd:|%s| RegisterForOnWriteableLocked called",
-                           GetName());
-    } else {
-      GPR_ASSERT(socket_type_ == SOCK_STREAM);
-      GRPC_CARES_TRACE_LOG(
-          "fd:|%s| RegisterForOnWriteableLocked called tcp_write_state_: %d",
-          GetName(), tcp_write_state_);
-    }
+  void RegisterForOnWriteableLocked(grpc_closure* write_closure) override {
+    GRPC_CARES_TRACE_LOG(
+        "fd:|%s| RegisterForOnWriteableLocked called connect_done_: %d "
+        "last_wsa_send_result_: %d",
+        GetName(), connect_done_, last_wsa_send_result_);
     GPR_ASSERT(write_closure_ == nullptr);
     write_closure_ = write_closure;
-    if (connect_done_) {
-      ContinueRegisterForOnWriteableLocked();
-    } else {
-      GPR_ASSERT(pending_continue_register_for_on_writeable_locked_ == false);
+    if (!connect_done_) {
+      GPR_ASSERT(!pending_continue_register_for_on_writeable_locked_);
       pending_continue_register_for_on_writeable_locked_ = true;
+      grpc_socket_notify_on_write(winsocket_, &on_tcp_connect_locked_);
```

Comment:
Ok, this seems very subtle. Did the issue happen when we are shutting down between an async TCP connect and its completion? What exactly happened? Did the `OnTcpConnect` callback get called because the poller was shutting down and causing crash or further failures?

Reply:
Yeah, that makes sense. Suggest add a comment here: `// Register an async OnTcpConnect callback here since we are guaranteed to hold a ref of the c-ares wrapper before write_closure_ is called. `.

Reply:
added the comment